Adding Mongodb Driver

meabed opened the pull request:
- Adding Mongodb Driver
- Update Readme to include Mongodb

meabed committed:
- Adding Mongodb Driver
- Update Readme to include Mongodb

meabed commented:
Awesome project Thank you.
Small contribution to the project, let me know if it need any changes or fixes.

meabed committed:
Fix tests, callbacks, binary

meabed committed:
Fix tests, callbacks, binary

meabed committed:
Fix tests, callbacks, binary

meabed commented:
![screen shot 2017-12-27 at 11 47 34 pm](https://user-images.githubusercontent.com/45731/34391676-b99c0092-eb60-11e7-9297-2ac1cae15298.png)

meabed committed:
Fix tests, callbacks, binary

meabed committed:
Fix tests, callbacks, binary

jhuckaby commented:
Thank you for this!  I have never used MongoDB before, so I will have to learn a bit about it and install a local version, so I can test this.  If I merge this PR, I have to be able to support it going forward, and for that I have to understand how Mongo works.

I'm a bit concerned about converting all binary records to base64 -- is this really the only way to do it with Mongo?  That seems extremely wasteful.  I have applications that store 100MB binary records, for example.

Why are we creating an index on every startup?  Isn't this incorrect?  I am referring to this line:

```js
self.collection.createIndex({key: 1}, function (err, result) {
```

Maybe I just don't understand MongoDB, but it seems like the index should only need to be created on the collection **one time only**, when the collection is first created.

Small typo in your docs addition for `serialize` here:

```
When set to `true` (the default), this is left up to Mongodb to handle.
```

This logic is reversed.  When `serialize` is false, it sends the object directly to Mongo for it to do the serialization.  When `serialize` is true, **we** perform the serialization in Node.js first.

I might want to change the name of the module from `Mongodb ` to `MongoDB`, to better match their branding.

Thanks again!

meabed committed:
Fix tests, callbacks, binary

meabed commented:
Thanks @jhuckaby for the review :) 

No worries get sometime to use mongodb =)

- Binary conversation is fixed :) it store files as binary data Mongo collection file size limit 16MB [more](https://docs.mongodb.com/manual/reference/limits/#Sharding-Existing-Collection-Data-Size).

- To address the file size limits, I have added **GridFS** for large object / files storage in MongoDB [more](https://docs.mongodb.com/manual/core/gridfs/)

- createIndex to be able to do fast lookups 
`Creates an index on the specified field if the index does not already exist ` [more](https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/).

- serialize is fixed.

- Engine name changed to `MongoDB `

- I have added same test file for mongodb with configuration and ran both tests below.

Thanks 👍 let me know if you got chance to test it :) 

<img width="959" alt="screen shot 2017-12-31 at 3 14 04 am" src="https://user-images.githubusercontent.com/45731/34457882-b325ea5a-edd8-11e7-8a94-8d6ee54c8115.png">

The reason this driver is helpful is MongoDB is very popular and its 1 minute to get it up and running in docker container!

I have tried the Couchbase configuration but it took me ages to get it up :) and yet couldn't figure out many things :) must be me because i'm know to it, but nevertheless the more drivers the better :)

meabed committed:
Update callbacks, cleanup

meabed committed:
Fix tests, callbacks, binary

meabed commented:
ping :)

meabed commented:
@jhuckaby have you had a chance to look the PR :) anything needed i can do to help getting it tested and merged?

Thanks

jhuckaby commented:
Hello.  I am very sorry, but I have not had time to test this, and am not ready to merge the PR yet.  I have never used MongoDB in my entire life, so I am completely ignorant to it.  If I am to merge this PR in with the base code, I have to understand how to install it, and use it, and test it.  This will take some time.  I apologize, but I will get to this as soon as I can.  Thank you again for the great PR!

meabed commented:
no worries :) let me know if i could be in any help!
